### PR TITLE
Allow passing the selected surface to the Dev Console as a search param

### PR DESCRIPTION
**Pull request by vividviolet**
Related to https://github.com/Shopify/shopify-cli-extensions/issues/306

### Changes

Allow passing the selected surface to the Dev Console as a search param
- Filter extension selected surface for all events coming from the Dev Server

Manually set the surface for CLI 3.0 extensions to prevent them from being set to "admin" by default

### Tophat

1. Pull this branch and run `make clobber; make bootstrap`
2. Run `make run serve testdata/extension.config.yml`
3. Open http://localhost:8000?surface=admin
4. Verify that you only see Product Subscription extensions

image

### CODE REVIEW

File: packages/dev-console-app/package.json [hunk 1]
```diff
@@ -33,7 +33,7 @@
     "sass": "^1.42.1",
     "ts-jest": "^27.0.5",
     "ts-node": "^10.2.1",
-    "typescript": "^4.3.2",
+    "typescript": "^4.6.4",
```
Comment (vividviolet) [inline packages/dev-console-app/package.json:36]: This is needed to better support union types

File: packages/ui-extensions-server-kit/src/ExtensionServerClient/ExtensionServerClient.ts [hunk 2]
```diff
@@ -160,12 +162,21 @@ export class ExtensionServerClient implements ExtensionServer.Client {
...
       ...options.connection,
       ...newOptions.connection,
     },
-  };
+  });
+}
+
+function getValidatedOptions<TOptions extends DeepPartial<ExtensionServer.Options>>(
+  options: TOptions,
+): TOptions {
```
Comment (olavoasantos) [inline packages/ui-extensions-server-kit/src/ExtensionServerClient/ExtensionServerClient.ts:177]: I think this should return `ExtensionServer.Options`, no?

File: packages/ui-extensions-server-kit/src/ExtensionServerClient/ExtensionServerClient.ts [hunk 3]
```diff
@@ -20,14 +21,14 @@ export class ExtensionServerClient implements ExtensionServer.Client {
...
-    this.options = {
+    this.options = getValidatedOptions({
       ...options,
       connection: {
         automaticConnect: true,
         protocols: [],
         ...(options.connection ?? {}),
       },
-    } as ExtensionServer.Options;
+    }) as ExtensionServer.Options;
```
Comment (olavoasantos) [inline packages/ui-extensions-server-kit/src/ExtensionServerClient/ExtensionServerClient.ts:31]: Could we do something like:
```ts
this.options = getValidatedOptions(mergeOptions(DEFAULT_OPTIONS, options));
```

File: packages/ui-extensions-server-kit/src/utilities/isValidSurface.ts [hunk 4]
```diff
@@ -0,0 +1,5 @@
+import type {Surface} from '../ExtensionServerClient';
+
+export function isValidSurface(surface: any): surface is Surface {
+  return (surface && surface === 'admin') || surface === 'checkout' || surface === 'post-purchase';
```
Comment (olavoasantos) [inline packages/ui-extensions-server-kit/src/utilities/isValidSurface.ts:4]: NABD... maybe we could have a constant defining the available surfaces:
```ts
const AVAILABLE_SURFACES = ['admin', 'checkout', 'post-checkout'];

export function isValidSurface(surface: any): surface is Surface {
  return surface && AVAILABLE_SURFACES.includes(surface);
}
```

File: packages/ui-extensions-server-kit/src/ExtensionServerClient/ExtensionServerClient.ts [hunk 5]
```diff
@@ -160,12 +162,21 @@ export class ExtensionServerClient implements ExtensionServer.Client {
 }
 
 function mergeOptions(options: ExtensionServer.Options, newOptions: ExtensionServer.Options) {
```
[the remark this answers is not in the document]
Comment (vividviolet) [reply packages/ui-extensions-server-kit/src/ExtensionServerClient/ExtensionServerClient.ts:164]: It doesn't work - we can't take in `DeepPartial<ExtensionServer.Options>` without it complaining that certain values are missing. I'd rather not fix this in this PR as it's not critical

File: packages/ui-extensions-server-kit/src/ExtensionServerClient/ExtensionServerClient.ts (hunk 3 above)
Comment (vividviolet) [reply packages/ui-extensions-server-kit/src/ExtensionServerClient/ExtensionServerClient.ts:31]: `mergeOptions` already calls `getValidatedOptions` internally

Comment (vividviolet) [reply packages/ui-extensions-server-kit/src/ExtensionServerClient/ExtensionServerClient.ts:31]: I can't use mergeOptions here cause it complains of missing props on `protocols` - again, I'd rather not fix it in this PR as it's unrelated to the changes

File: packages/ui-extensions-server-kit/src/ExtensionServerClient/ExtensionServerClient.ts (hunk 2 above)
Comment (vividviolet) [reply packages/ui-extensions-server-kit/src/ExtensionServerClient/ExtensionServerClient.ts:177]: oops 🙃

File: packages/ui-extensions-server-kit/src/ExtensionServerClient/ExtensionServerClient.test.ts [hunk 6]
```diff
@@ -19,6 +24,16 @@ describe('ExtensionServerClient', () => {
...
+  // beforeEach(() => {
+  //   // eslint-disable-next-line jest/prefer-spy-on
+  //   global.fetch = jest.fn().mockImplementationOnce(() =>
+  //     Promise.resolve({
+  //       status: 400,
+  //       json: () => Promise.resolve({success: false, error: 'Something bad happened'}),
+  //     }),
+  //   );
+  // });
+
```
[the remark this answers is not in the document]
Comment (vividviolet) [reply packages/ui-extensions-server-kit/src/ExtensionServerClient/ExtensionServerClient.test.ts:36]: oops 🙃

File: packages/ui-extensions-server-kit/src/ExtensionServerClient/ExtensionServerClient.ts (hunk 2 above)
Comment (vividviolet) [reply packages/ui-extensions-server-kit/src/ExtensionServerClient/ExtensionServerClient.ts:177]: This is not possible until the issue with `DeepPartial` is fixed - it complains about setting `protocols`

**Review by olavoasantos — APPROVED** [review]
Looks good! 👍

**Pull request merged**